File: lib/src/analyzer/board_space.cc
// Copyright (c) 2020 The DashQL Authors

#include "dashql/analyzer/board_space.h"

#include <iostream>

static constexpr uint32_t CELL_MASK_SHIFT = 3;
static constexpr uint32_t COLUMNS_PER_ROW = 12;
static constexpr uint32_t MAX_ROWS = 100;
static constexpr uint32_t MAX_HEIGHT = 40;
static constexpr uint32_t DEFAULT_WIDTH = 12;
static constexpr uint32_t DEFAULT_HEIGHT = 4;

namespace dashql {

namespace {

inline bool isSet(const std::vector<uint8_t>& cells, size_t row, size_t col) {
    auto i = row * COLUMNS_PER_ROW + col;
    auto& block = cells[i >> CELL_MASK_SHIFT];
    auto local = i & ((1 << CELL_MASK_SHIFT) - 1);
    return (block & (1 << local)) != 0;
}

inline void set(std::vector<uint8_t>& cells, size_t row, size_t col) {
    auto i = row * COLUMNS_PER_ROW + col;
    auto& block = cells[i >> CELL_MASK_SHIFT];
    auto local = i & ((1 << CELL_MASK_SHIFT) - 1);
    block |= (1 << local);
}

inline void unset(std::vector<uint8_t>& cells, size_t row, size_t col) {
    auto i = row * COLUMNS_PER_ROW + col;
    auto& block = cells[i >> CELL_MASK_SHIFT];
    auto local = i & ((1 << CELL_MASK_SHIFT) - 1);
    block &= ~(1 << local);
};

}  // namespace

/// Constructor
BoardSpace::BoardSpace() : cells() {}

/// Allocate a position
BoardPosition BoardSpace::Allocate(BoardPosition pref) {
    // Make sure that at least pref.row + pref.height rows are allocated
    pref.width = std::min(pref.width, COLUMNS_PER_ROW);
    pref.height = std::min(pref.height, MAX_HEIGHT);
    pref.width = pref.width == 0 ? DEFAULT_WIDTH : pref.width;
    pref.height = pref.height == 0 ? DEFAULT_HEIGHT : pref.height;
    pref.row = std::min(pref.row, MAX_ROWS - pref.height);
    pref.column = std::min(pref.column, COLUMNS_PER_ROW - pref.width);

    // Resize the block mask
    auto max_columns = std::min(COLUMNS_PER_ROW - pref.width + 1, COLUMNS_PER_ROW);
    auto max_rows = pref.row + pref.height;
    auto current_rows = (cells.size() << CELL_MASK_SHIFT) / COLUMNS_PER_ROW;
    if (current_rows < max_rows) {
        cells.resize(((max_rows * COLUMNS_PER_ROW) >> CELL_MASK_SHIFT) + 1, 0);
    }

    // Naive brute-force space allocation.
    // We could be smarter here but it's very likely not necessary.
    uint32_t row = 0, col = 0;
    for (row = pref.row; row < max_rows; ++row) {
        for (col = (row == pref.row) ? pref.column : 0; col < max_columns; ++col) {
            bool qualifies = true;
            for (auto r = row; r < (row + pref.height); ++r) {
                for (auto c = col; c < (col + pref.width); ++c) {
                    qualifies &= !isSet(cells, r, c);
                }
            }
            if (qualifies) {
                goto found_space;
            }
        }
    }

    // Could not allocate the block?
    // Resize the buffer.
    max_rows = row + pref.height;
    cells.resize(((max_rows * COLUMNS_PER_ROW) >> CELL_MASK_SHIFT) + 1, 0);

found_space:
    // Mark the bits as occupied
    for (size_t r = row; r < (row + pref.height); ++r) {
        for (size_t c = col; c < (col + pref.width); ++c) {
            set(cells, r, c);
        }
    }

    // Return position
    return {
        .width = pref.width,
        .height = pref.height,
        .row = row,
        .column = col,
    };
}

}  // namespace dashql

File: dvms/src/analyzer/board_space.rs
const CELL_MASK_SHIFT: usize = 3;
const COLUMNS_PER_ROW: usize = 12;
const MAX_ROWS: usize = 100;
const MAX_HEIGHT: usize = 40;
const DEFAULT_WIDTH: usize = 12;
const DEFAULT_HEIGHT: usize = 4;

#[derive(Default)]
pub struct BoardPosition {
    pub row: usize,
    pub column: usize,
    pub width: usize,
    pub height: usize,
}

#[derive(Default)]
pub struct BoardSpace {
    /// The cell bitmask. 1 = occupied, 0 = free.
    pub cells: Vec<u8>,
}

impl BoardSpace {
    pub fn is_set(&self, row: usize, col: usize) -> bool {
        let i = row * COLUMNS_PER_ROW + col;
        let block = &self.cells[i >> CELL_MASK_SHIFT];
        let local = i & ((1 << CELL_MASK_SHIFT) - 1);
        return (block & (1 << local)) != 0;
    }

    pub fn set(&mut self, row: usize, col: usize) {
        let i = row * COLUMNS_PER_ROW + col;
        let local = i & ((1 << CELL_MASK_SHIFT) - 1);
        self.cells[i >> CELL_MASK_SHIFT] |= (1 << local) as u8;
    }

    pub fn unset(&mut self, row: usize, col: usize) {
        let i = row * COLUMNS_PER_ROW + col;
        let local = i & ((1 << CELL_MASK_SHIFT) - 1);
        self.cells[i >> CELL_MASK_SHIFT] &= !(1 << local) as u8;
    }

    /// Allocate a board position
    pub fn allocate(&mut self, mut pref: BoardPosition) -> BoardPosition {
        // Make sure that at least pref.row + pref.height rows are allocated
        pref.width = pref.width.min(COLUMNS_PER_ROW);
        pref.height = pref.height.min(MAX_HEIGHT);
        pref.width = if pref.width == 0 { DEFAULT_WIDTH } else { pref.width };
        pref.height = if pref.height == 0 { DEFAULT_HEIGHT } else { pref.height };
        pref.row = pref.row.min(MAX_ROWS - pref.height);
        pref.column = pref.column.min(COLUMNS_PER_ROW - pref.width);

        // Resize the block mask
        let column_candidates = COLUMNS_PER_ROW.min(COLUMNS_PER_ROW - pref.width + 1);
        let required_rows = pref.row + pref.height;
        let mut row_count = (self.cells.len() << CELL_MASK_SHIFT) / COLUMNS_PER_ROW;
        if row_count < required_rows {
            self.cells
                .resize(((required_rows * COLUMNS_PER_ROW) >> CELL_MASK_SHIFT) + 1, 0);
            row_count = required_rows;
        }

        // Brute-force space allocation.
        // We could be smarter here but it's very likely not necessary.
        loop {
            // Naively check every origin candidate
            for row in pref.row..(row_count - pref.height) {
                for col in (if row == pref.row { pref.column } else { 0 })..column_candidates {
                    // Check if all cells are free
                    let mut qualifies = true;
                    for r in row..(row + pref.height) {
                        for c in col..(col + pref.width) {
                            qualifies &= !self.is_set(r, c);
                        }
                    }
                    // Does not qualify?
                    if !qualifies {
                        continue;
                    }
                    // Mark the cells as occupied
                    for r in row..(row + pref.height) {
                        for c in col..(col + pref.width) {
                            self.set(r, c);
                        }
                    }
                    // Return the position
                    return BoardPosition {
                        row,
                        column: col,
                        width: pref.width,
                        height: pref.height,
                    };
                }
            }
            // Could not allocate the block?
            // Resize the buffer.
            row_count += pref.height;
            self.cells
                .resize(((row_count * COLUMNS_PER_ROW) >> CELL_MASK_SHIFT) + 1, 0)
        }
    }
}
